pkg/core/region.go

```diff
@@ -1368,6 +1368,57 @@ func (r *RegionsInfo) GetStoreRegions(storeID uint64) []*RegionInfo {
 	return regions
 }
 
+// SubTreeRegionType is the type of sub tree region.
+type SubTreeRegionType int
+
+var (
+	// AllInSubTree is all sub trees.
+	AllInSubTree SubTreeRegionType = 0
+	// LeaderInSubTree is the leader sub tree.
+	LeaderInSubTree SubTreeRegionType = 1
+	// FollowerInSubTree is the follower sub tree.
+	FollowerInSubTree SubTreeRegionType = 2
+	// LearnerInSubTree is the learner sub tree.
+	LearnerInSubTree SubTreeRegionType = 3
+	// WitnessInSubTree is the witness sub tree.
+	WitnessInSubTree SubTreeRegionType = 4
+	// PendingPeerInSubTree is the pending peer sub tree.
+	PendingPeerInSubTree SubTreeRegionType = 5
+)
+
+// GetStoreRegions gets all RegionInfo with a given storeID
+func (r *RegionsInfo) GetStoreRegionsByTypeInSubTree(storeID uint64, typ SubTreeRegionType) []*RegionInfo {
+	r.st.RLock()
+	var regions []*RegionInfo
+	switch typ {
+	case LeaderInSubTree:
+		if leaders, ok := r.leaders[storeID]; ok {
+			regions = leaders.scanRanges()
+		}
+	case FollowerInSubTree:
+		if followers, ok := r.followers[storeID]; ok {
+			regions = followers.scanRanges()
+		}
+	case LearnerInSubTree:
+		if learners, ok := r.learners[storeID]; ok {
+			regions = learners.scanRanges()
+		}
+	case WitnessInSubTree:
+		if witnesses, ok := r.witnesses[storeID]; ok {
+			regions = witnesses.scanRanges()
+		}
+	case PendingPeerInSubTree:
+		if pendingPeers, ok := r.pendingPeers[storeID]; ok {
+			regions = pendingPeers.scanRanges()
+		}
+	case AllInSubTree:
+		r.st.RUnlock()
+		return r.GetStoreRegions(storeID)
+	}
```

Comment:
Maybe add a fallback logic to avoid forgetting to update the logic here when adding more types in the future.